### DISCUSSION TITLE
Run feature test in parallel does not respect "maxParallelThreads" in xUnit

### DISCUSSION THREAD

**Discussion by llbonet14** [reqnroll/Reqnroll#328] [Q&A] [answered]
I'm migrating my specflow test project to reqnroll, I've been able to fix some issues I've found, but there's one issue I'm not able to fix. I used to run the feature files in parallel in specflow where the maxParallelThreads set the maximum number of feature files to run in parallel, so if set to 2 there where 2 drivers running all the scenarios from that feature file and no other feature file started until one was finished. But what currently happens is that at the start the max threats are respected and only that number of feature files are run, but as soon as one of the scenarios has finished (even though there are many more in that feature) another feature has started and it continues without respecting the maxParallelThreads. This is my xunit config file (which worked with specflow)

`{
  "$schema": "https://xunit.net/schema/current/xunit.runner.schema.json",
  "parallelizeAssembly": false,
  "parallelizeTestCollections": true,
  "maxParallelThreads": 3,
  "parallelAlgorithm": "aggressive",
  "diagnosticMessages": true,
  "internalDiagnosticMessages": true
}` 

This is how it starts
![image](https://github.com/user-attachments/assets/b25a5d39-3fe9-4371-ab0f-818e95fe647e)

This is how continues
![image](https://github.com/user-attachments/assets/d0de724d-2f05-428f-bbd4-57fa7eeb3a58)

This is my VS Test explorer with the feature files
![image](https://github.com/user-attachments/assets/13d98a05-9bfc-484d-b0ae-58ec9e06ab32)

**Comment by gasparnagy** [comment]
I have tried to reproduce this with our basic sample project: https://github.com/reqnroll/Reqnroll.ExploratoryTestProjects/tree/main/ParallelExecution/ReqnrollCalculator.Specs.XUnitFw, but could not.

In that sample there are 3 feature files and if I set the "maxParallelThreads" to "2" from json (or from code) it completes the two features is starts with and only after it continues with the third one. 

Could you please try to reproduce it on a minimalistic sample project that you can share? 

In addition to that, you can try to change the config from code, instead of json to check if it has any impact (see sample [here](https://github.com/reqnroll/Reqnroll.ExploratoryTestProjects/blob/main/ParallelExecution/ReqnrollCalculator.Specs.XUnitFw/ParallelConfig.cs)).

Also you can try to make some logging to a file, like what I do [here](https://github.com/reqnroll/Reqnroll.ExploratoryTestProjects/blob/main/ParallelExecution/ReqnrollCalculator.Specs.StepDefinitions/LoggingHooks.cs). This produces a log file where you can track the execution order. In my example it produced:

```
00:00:00.0008251 -: BeforeTestRun
00:00:00.2299334 2:   BeforeFeature: Addition2
00:00:00.2299493 3:   BeforeFeature: Addition3
00:00:00.2459359 2:     BeforeScenario: Addition2 / Add two numbers
00:00:00.2503100 3:     BeforeScenario: Addition3 / Add zeros / #1,0,0,0
00:00:00.6419231 3:     AfterScenario:  Addition3 / Add zeros / #1,0,0,0
00:00:00.6478406 2:     AfterScenario:  Addition2 / Add two numbers
00:00:00.6569974 3:     BeforeScenario: Addition3 / Add zeros / #2,0,42,42
00:00:00.6636027 2:     BeforeScenario: Addition2 / Add zeros / #3,42,0,42
00:00:00.9963253 3:     AfterScenario:  Addition3 / Add zeros / #2,0,42,42
00:00:01.0169510 3:     BeforeScenario: Addition3 / Add zeros / #3,42,0,42
00:00:01.0352450 2:     AfterScenario:  Addition2 / Add zeros / #3,42,0,42
00:00:01.0543524 2:     BeforeScenario: Addition2 / Add zeros / #2,0,42,42
00:00:01.3837901 2:     AfterScenario:  Addition2 / Add zeros / #2,0,42,42
00:00:01.3862067 2:     BeforeScenario: Addition2 / Add zeros / #1,0,0,0
00:00:01.3920720 3:     AfterScenario:  Addition3 / Add zeros / #3,42,0,42
00:00:01.3963330 3:     BeforeScenario: Addition3 / Add negatives
00:00:01.7506050 2:     AfterScenario:  Addition2 / Add zeros / #1,0,0,0
00:00:01.7536774 2:     BeforeScenario: Addition2 / Addition is commutative
00:00:01.7612151 3:     AfterScenario:  Addition3 / Add negatives
00:00:01.7644206 3:     BeforeScenario: Addition3 / Addition is commutative
00:00:02.0936738 2:     AfterScenario:  Addition2 / Addition is commutative
00:00:02.0966115 2:     BeforeScenario: Addition2 / Add negatives
00:00:02.1375964 3:     AfterScenario:  Addition3 / Addition is commutative
00:00:02.1582099 3:     BeforeScenario: Addition3 / Add two numbers
00:00:02.4439418 2:     AfterScenario:  Addition2 / Add negatives
00:00:02.4536669 2:   AfterFeature:  Addition2
00:00:02.4554467 2:   BeforeFeature: Addition1
00:00:02.4581570 2:     BeforeScenario: Addition1 / Add two numbers
00:00:02.6434320 3:     AfterScenario:  Addition3 / Add two numbers
00:00:02.8746116 2:     AfterScenario:  Addition1 / Add two numbers
00:00:02.8842305 2:     BeforeScenario: Addition1 / Add zeros / #2,0,42,42
00:00:03.2621823 2:     AfterScenario:  Addition1 / Add zeros / #2,0,42,42
00:00:03.2720470 2:     BeforeScenario: Addition1 / Add zeros / #1,0,0,0
00:00:03.6359258 2:     AfterScenario:  Addition1 / Add zeros / #1,0,0,0
00:00:03.6374110 2:     BeforeScenario: Addition1 / Add zeros / #3,42,0,42
00:00:03.9701528 2:     AfterScenario:  Addition1 / Add zeros / #3,42,0,42
00:00:03.9744614 2:     BeforeScenario: Addition1 / Addition is commutative
00:00:04.3234573 2:     AfterScenario:  Addition1 / Addition is commutative
00:00:04.3370615 2:     BeforeScenario: Addition1 / Add negatives
00:00:04.6828261 2:     AfterScenario:  Addition1 / Add negatives
00:00:04.7363695 3:   AfterFeature:  Addition3
00:00:04.7461307 2:   AfterFeature:  Addition1
00:00:04.7492284 -: AfterTestRun
```

**Comment by llbonet14** [comment]
Hi @gasparnagy thanks for your answer, I tried to log the order of the test, in this case I run 3 feature files with the maxParallelThreads set to 2 also, the Auth has 5 tests scenario outlines for a total  of  tests, the Products has 7 tests scenario  outlines for a total of 7 tests and the FileProcessor has 3 scenario for a total of 3 tests and also I have a [ScenarioDependencies] tag.

Looking at the logs I see that it starts running 2 tests but then it continues with 3 and the after feature log is run at the end in all iterations I've tried.

```
00:00:00.0002871 -: BeforeTestRun
00:00:17.6643495 2:   BeforeFeature: Auth
00:00:17.6643566 3:   BeforeFeature: Products
00:00:25.6791077 3:     BeforeScenario: Products / Create and validate custom attributes / admin
00:00:25.7362900 2:     BeforeScenario: Auth / Log in with a 2FA user enabled / 2FA user
00:00:48.7973879 4:   BeforeFeature: FileProcessor
00:00:57.7410669 4:     BeforeScenario: FileProcessor / Create invoices through the FileProcessor with expiration date
00:01:16.6533716 3:     AfterScenario:  Products / Create and validate custom attributes / admin
00:01:32.2506192 3:     BeforeScenario: Products / Create a product / admin
00:02:03.2116172 2:     AfterScenario:  Auth / Log in with a 2FA user enabled / 2FA user
00:02:17.6316489 2:     BeforeScenario: Auth / Login logout / simple
00:02:22.1576942 2:     AfterScenario:  Auth / Login logout / simple
00:02:23.7869757 4:     AfterScenario:  FileProcessor / Create invoices through the FileProcessor with expiration date
00:02:37.4568522 2:     BeforeScenario: Auth / Login logout / department
00:02:41.0109705 4:     BeforeScenario: FileProcessor / Create invoices through the FileProcessor with ING merchant
00:02:42.2354299 2:     AfterScenario:  Auth / Login logout / department
00:02:58.0296899 2:     BeforeScenario: Auth / Login logout / admin
00:03:02.4199249 2:     AfterScenario:  Auth / Login logout / admin
00:03:16.9231804 2:     BeforeScenario: Auth / Login logout / main
00:03:21.4777398 2:     AfterScenario:  Auth / Login logout / main
00:03:22.9356419 4:     AfterScenario:  FileProcessor / Create invoices through the FileProcessor with ING merchant
00:03:43.6021888 2:     BeforeScenario: Auth / Log in with a disabled user / my
00:03:43.7442250 4:     BeforeScenario: FileProcessor / Create invoices through the FileProcessor without expiration date
00:03:45.0256351 2:     AfterScenario:  Auth / Log in with a disabled user / my
00:04:01.7847151 2:     BeforeScenario: Auth / Send forgot password email for a non-existing user / a non-existing
00:04:03.1138584 2:     AfterScenario:  Auth / Send forgot password email for a non-existing user / a non-existing
00:04:19.6172551 2:     BeforeScenario: Auth / Change user password / an existing
00:04:26.5783909 2:     AfterScenario:  Auth / Change user password / an existing
00:04:34.5740775 3:     AfterScenario:  Products / Create a product / admin
00:04:51.7097619 3:     BeforeScenario: Products / Create and associate an external product only / admin
00:05:22.5727724 3:     AfterScenario:  Products / Create and associate an external product only / admin
00:05:23.2262543 4:     AfterScenario:  FileProcessor / Create invoices through the FileProcessor without expiration date
00:05:38.5427646 3:     BeforeScenario: Products / Associate a product to a new organization / admin
00:05:53.9866929 3:     AfterScenario:  Products / Associate a product to a new organization / admin
00:06:08.8750465 3:     BeforeScenario: Products / Create and edit custom attributes / admin
00:06:50.7549811 3:     AfterScenario:  Products / Create and edit custom attributes / admin
00:07:06.5938315 3:     BeforeScenario: Products / Create and associate a product / admin
00:07:58.8089218 3:     AfterScenario:  Products / Create and associate a product / admin
00:08:15.2883280 3:     BeforeScenario: Products / Associate a product to a new department / admin
00:08:30.4144845 3:     AfterScenario:  Products / Associate a product to a new department / admin
00:08:38.4490768 2:   AfterFeature:  Auth
00:08:38.4497821 4:   AfterFeature:  FileProcessor
00:08:38.4500424 3:   AfterFeature:  Products
```
![image](https://github.com/user-attachments/assets/4960cde2-9923-446d-bd16-86f1e2fbe4f5)

**Reply by gasparnagy** [reply]
@llbonet14 Are you sure you have configured xUnit worker counts correcty? Because from your trace, it seems that there are 3 workers running in parallel (the number after the timestamp is the worker id):

```
00:00:48.7973879 4:   BeforeFeature: FileProcessor
00:00:57.7410669 4:     BeforeScenario: FileProcessor / Create invoices through the FileProcessor with expiration date
00:01:16.6533716 3:     AfterScenario:  Products / Create and validate custom attributes / admin
00:01:32.2506192 3:     BeforeScenario: Products / Create a product / admin
00:02:03.2116172 2:     AfterScenario:  Auth / Log in with a 2FA user enabled / 2FA user
00:02:17.6316489 2:     BeforeScenario: Auth / Login logout / simple
```

So it seems that worker 4 deals with `FileProcessor` worker 3 with `Products` and worker 2 with `Auth`.

Please try to set the config from code (see [example](https://github.com/reqnroll/Reqnroll.ExploratoryTestProjects/blob/main/ParallelExecution/ReqnrollCalculator.Specs.XUnitFw/ParallelConfig.cs)) to make sure the problem is not with the xUnit JSON config.

**Comment by llbonet14** [comment]
I use this file `xunit.runner.json` with this values:

`{
  "$schema": "https://xunit.net/schema/current/xunit.runner.schema.json",
  "parallelizeAssembly": false,
  "parallelizeTestCollections": true,
  "maxParallelThreads": 2,
  "parallelAlgorithm": "aggressive",
  "diagnosticMessages": true,
  "internalDiagnosticMessages": true
}`

And this is the configuration I use with specflow, the strange thing is that worker 4 does not start until one test of other workers have finished. As you can see worker 2 and 3 start at the same time and after 48 seconds the other worker starts.

Also the all the AfterFeature are run at the same time after all test from all featureFiles have finished, but some features have finished before

**Reply by gasparnagy** [reply]
> worker 4 does not start until one test of other workers have finished

This is not true. If you check the beginning of your log, you can see that the first "AfterScenario" (so the first finished test) is at 1:16, while worker 4 starts already at 0:48.

```
00:00:00.0002871 -: BeforeTestRun
00:00:17.6643495 2:   BeforeFeature: Auth
00:00:17.6643566 3:   BeforeFeature: Products
00:00:25.6791077 3:     BeforeScenario: Products / Create and validate custom attributes / admin
00:00:25.7362900 2:     BeforeScenario: Auth / Log in with a 2FA user enabled / 2FA user
00:00:48.7973879 4:   BeforeFeature: FileProcessor
00:00:57.7410669 4:     BeforeScenario: FileProcessor / Create invoices through the FileProcessor with expiration date
00:01:16.6533716 3:     AfterScenario:  Products / Create and validate custom attributes / admin
```

> Also the all the AfterFeature are run at the same time after all test from all featureFiles have finished, but some features have finished before

This is a behavior that Reqnroll introduced. It delays the execution of the AfterFeature until a test from another feature has been scheduled to the same worker. So this is by design.

> I use this file xunit.runner.json with this values

I don't know. From your log it seems that there are 3 workers running: for example at 1:00 of your execution time the scenarios

* Products / Create and validate custom attributes / admin
* Auth / Log in with a 2FA user enabled / 2FA user
* FileProcessor / Create invoices through the FileProcessor with expiration date

have all started and none of them finished yet. As you see my trace above, it worked for me, so there must be something special in your setup that causes this. You should somehow try to figure out what it is, so that we can help. This is why I suggested to test the code-based configuration, but maybe there is something else too.

**Comment by llbonet14** [comment]
I downloaded your test project `Reqnroll.ExploratoryTestProjects/ParallelExecution` (https://github.com/reqnroll/Reqnroll.ExploratoryTestProjects/tree/main/ParallelExecution) and updated all xunit and nugets dependencies to latest versions, added more feature files, set the max parallel to 2 and log the results I think the max count is not respected but maybe you see that they so, here are the logs I have a local branch with the changes

```
00:00:00.0002783 -: BeforeTestRun
00:00:00.0351080 6:   BeforeFeature: Addition6
00:00:00.0351411 3:   BeforeFeature: Addition3
00:00:00.0351500 7:   BeforeFeature: Addition7
00:00:00.0351547 2:   BeforeFeature: Addition1
00:00:00.0351561 8:   BeforeFeature: Addition2
00:00:00.0351611 4:   BeforeFeature: Addition5
00:00:00.0351822 5:   BeforeFeature: Addition4
00:00:00.0448249 6:     BeforeScenario: Addition6 / Add two numbers
00:00:00.0448250 7:     BeforeScenario: Addition7 / Add two numbers
00:00:00.0448249 8:     BeforeScenario: Addition2 / Add two numbers
00:00:00.0448311 2:     BeforeScenario: Addition1 / Add two numbers
00:00:00.0448453 3:     BeforeScenario: Addition3 / Add negatives
00:00:00.0465137 4:     BeforeScenario: Addition5 / Add zeros / #1,0,0,0
00:00:00.0465167 5:     BeforeScenario: Addition4 / Add zeros / #1,0,0,0
00:00:00.3727616 8:     AfterScenario:  Addition2 / Add two numbers
00:00:00.3789606 8:     BeforeScenario: Addition2 / Add zeros / #1,0,0,0
00:00:00.3852621 2:     AfterScenario:  Addition1 / Add two numbers
00:00:00.3855892 5:     AfterScenario:  Addition4 / Add zeros / #1,0,0,0
00:00:00.3854400 3:     AfterScenario:  Addition3 / Add negatives
00:00:00.3864160 5:     BeforeScenario: Addition4 / Add zeros / #2,0,42,42
00:00:00.3869709 2:     BeforeScenario: Addition1 / Add zeros / #2,0,42,42
00:00:00.3871195 3:     BeforeScenario: Addition3 / Addition is commutative
00:00:00.4004457 7:     AfterScenario:  Addition7 / Add two numbers
00:00:00.4005079 6:     AfterScenario:  Addition6 / Add two numbers
00:00:00.4008502 4:     AfterScenario:  Addition5 / Add zeros / #1,0,0,0
00:00:00.4020967 7:     BeforeScenario: Addition7 / Add zeros / #3,42,0,42
00:00:00.4027196 6:     BeforeScenario: Addition6 / Add zeros / #3,42,0,42
00:00:00.4028504 4:     BeforeScenario: Addition5 / Add zeros / #3,42,0,42
00:00:00.7099161 5:     AfterScenario:  Addition4 / Add zeros / #2,0,42,42
00:00:00.7099178 8:     AfterScenario:  Addition2 / Add zeros / #1,0,0,0
00:00:00.7099193 3:     AfterScenario:  Addition3 / Addition is commutative
00:00:00.7113348 8:     BeforeScenario: Addition2 / Add zeros / #2,0,42,42
00:00:00.7118109 5:     BeforeScenario: Addition4 / Add zeros / #3,42,0,42
00:00:00.7120661 3:     BeforeScenario: Addition3 / Add zeros / #1,0,0,0
00:00:00.7244713 4:     AfterScenario:  Addition5 / Add zeros / #3,42,0,42
00:00:00.7246105 7:     AfterScenario:  Addition7 / Add zeros / #3,42,0,42
00:00:00.7256413 4:     BeforeScenario: Addition5 / Add zeros / #2,0,42,42
00:00:00.7259758 7:     BeforeScenario: Addition7 / Add zeros / #2,0,42,42
00:00:00.7404837 6:     AfterScenario:  Addition6 / Add zeros / #3,42,0,42
00:00:00.7411496 2:     AfterScenario:  Addition1 / Add zeros / #2,0,42,42
00:00:00.7421121 6:     BeforeScenario: Addition6 / Add zeros / #1,0,0,0
00:00:00.7426100 2:     BeforeScenario: Addition1 / Add zeros / #3,42,0,42
00:00:01.0328675 7:     AfterScenario:  Addition7 / Add zeros / #2,0,42,42
00:00:01.0328819 8:     AfterScenario:  Addition2 / Add zeros / #2,0,42,42
00:00:01.0329015 5:     AfterScenario:  Addition4 / Add zeros / #3,42,0,42
00:00:01.0338052 7:     BeforeScenario: Addition7 / Add zeros / #1,0,0,0
00:00:01.0340212 8:     BeforeScenario: Addition2 / Add zeros / #3,42,0,42
00:00:01.0349079 5:     BeforeScenario: Addition4 / Add negatives
00:00:01.0486957 3:     AfterScenario:  Addition3 / Add zeros / #1,0,0,0
00:00:01.0498322 3:     BeforeScenario: Addition3 / Add zeros / #3,42,0,42
00:00:01.0646983 2:     AfterScenario:  Addition1 / Add zeros / #3,42,0,42
00:00:01.0654898 2:     BeforeScenario: Addition1 / Add zeros / #1,0,0,0
00:00:01.0802279 6:     AfterScenario:  Addition6 / Add zeros / #1,0,0,0
00:00:01.0802281 4:     AfterScenario:  Addition5 / Add zeros / #2,0,42,42
00:00:01.0810168 6:     BeforeScenario: Addition6 / Add zeros / #2,0,42,42
00:00:01.0816207 4:     BeforeScenario: Addition5 / Add negatives
00:00:01.3443852 7:     AfterScenario:  Addition7 / Add zeros / #1,0,0,0
00:00:01.3443852 8:     AfterScenario:  Addition2 / Add zeros / #3,42,0,42
00:00:01.3458309 8:     BeforeScenario: Addition2 / Addition is commutative
00:00:01.3458577 7:     BeforeScenario: Addition7 / Addition is commutative
00:00:01.3753333 5:     AfterScenario:  Addition4 / Add negatives
00:00:01.3766711 5:     BeforeScenario: Addition4 / Add two numbers
00:00:01.3908246 3:     AfterScenario:  Addition3 / Add zeros / #3,42,0,42
00:00:01.3908251 6:     AfterScenario:  Addition6 / Add zeros / #2,0,42,42
00:00:01.3915036 3:     BeforeScenario: Addition3 / Add zeros / #2,0,42,42
00:00:01.3924368 6:     BeforeScenario: Addition6 / Addition is commutative
00:00:01.4061063 2:     AfterScenario:  Addition1 / Add zeros / #1,0,0,0
00:00:01.4073333 2:     BeforeScenario: Addition1 / Addition is commutative
00:00:01.4223087 4:     AfterScenario:  Addition5 / Add negatives
00:00:01.4235364 4:     BeforeScenario: Addition5 / Addition is commutative
00:00:01.6548787 8:     AfterScenario:  Addition2 / Addition is commutative
00:00:01.6567494 8:     BeforeScenario: Addition2 / Add negatives
00:00:01.7011210 7:     AfterScenario:  Addition7 / Addition is commutative
00:00:01.7028244 7:     BeforeScenario: Addition7 / Add negatives
00:00:01.7171503 5:     AfterScenario:  Addition4 / Add two numbers
00:00:01.7171671 2:     AfterScenario:  Addition1 / Addition is commutative
00:00:01.7172501 6:     AfterScenario:  Addition6 / Addition is commutative
00:00:01.7187133 5:     BeforeScenario: Addition4 / Addition is commutative
00:00:01.7190158 2:     BeforeScenario: Addition1 / Add negatives
00:00:01.7205432 6:     BeforeScenario: Addition6 / Add negatives
00:00:01.7321243 3:     AfterScenario:  Addition3 / Add zeros / #2,0,42,42
00:00:01.7341687 3:     BeforeScenario: Addition3 / Add two numbers
00:00:01.7474875 4:     AfterScenario:  Addition5 / Addition is commutative
00:00:01.7493719 4:     BeforeScenario: Addition5 / Add two numbers
00:00:01.9960329 8:     AfterScenario:  Addition2 / Add negatives
00:00:02.0269142 5:     AfterScenario:  Addition4 / Addition is commutative
00:00:02.0581911 2:     AfterScenario:  Addition1 / Add negatives
00:00:02.0581912 7:     AfterScenario:  Addition7 / Add negatives
00:00:02.0742724 4:     AfterScenario:  Addition5 / Add two numbers
00:00:02.0742724 6:     AfterScenario:  Addition6 / Add negatives
00:00:02.0742724 3:     AfterScenario:  Addition3 / Add two numbers
00:00:02.0801698 7:   AfterFeature:  Addition7
00:00:02.0805650 4:   AfterFeature:  Addition5
00:00:02.0808115 6:   AfterFeature:  Addition6
00:00:02.0810560 3:   AfterFeature:  Addition3
00:00:02.0812856 5:   AfterFeature:  Addition4
00:00:02.0815093 2:   AfterFeature:  Addition1
00:00:02.0817205 8:   AfterFeature:  Addition2
00:00:02.0827576 -: AfterTestRun
```

**Comment by obligaron** [comment]
Could this be related to https://github.com/xunit/xunit/issues/2003?
Reqnroll itself always uses async APIs internally (see generated .feature.cs files). So the problem described in that issue could also be the cause here. 🤔

**Reply by gasparnagy** [reply]
That is a good spot @obligaron! @llbonet14 Could you please confirm that the changes you made on the sample project were related to async execution? Could you please just share all changes you made?

**Reply by llbonet14** [reply]
Hi @gasparnagy,
Here I have the changes but basically I updated the xunit nuget and add some more feature files seting the max parallel to 2. The feature files are copies from the existing ones just changing the names. I did not change any test logic, that's when I thought it could be related with xunit because they had some break changes in version 2.8 I had to adapt my project to continue working with the `"maxParallelThreads": "2"` I had to set `"parallelAlgorithm": "aggressive"` to make it work


![image](https://github.com/user-attachments/assets/afe6d57a-c00e-4f89-9a03-107b9314b745)
![image](https://github.com/user-attachments/assets/90f109a2-32d3-4f49-af18-3d8a7d8e7e8a)
![image](https://github.com/user-attachments/assets/d524ce52-da87-4e3a-a400-d0353fd77a22)

**Reply by llbonet14** [reply]
Hi @gasparnagy could you reproduce this behavior in your local project with the latest xunit versions?

**Reply by llbonet14** [reply]
I have migrated from xUnit to Nunit and using this assembly it works perfectly, so it's a xunit problem with the await async tests.

**Reply by ajeckmans** [reply]
No it is a misunderstanding of what that setting does. MaxParallelThreads controls how many test threads are started. But it never was about what the test thread actually does. It can run multiple tasks consecutively from different tests.

So if both test A and B have awaits, the following is a valid run scenario:
Thread#1: A_1, B_1, A_2, B_2

With two threads it could become something like the following (though unlikely):
Thread#1: A_1, B_2
Thread#2: B_1, A_2

See also: https://github.com/xunit/xunit/discussions/2670
Xunit simply has no feature available to control the number of _started_ tests (what you guys are looking for), only the amount of _actively running_ tests

You could very well end up with more started tests (and features) that there are test threads configured, but that is more of an edge case.

**Reply by DerAlbertCom** [reply, marked as answer]
Since Xunit 2.8, Xunit has the `conservative` mode for https://xunit.net/docs/configuration-files#parallelAlgorithm, which then ensures that only such much test are running concurrently  which are configured as max threads. That works.

However, this does not work with Reqnroll.Xunit. Because the Regnroll runner which is registered with an Assembly Attribute which Regnroll creates does not know about this, and does not switch the mode. I would assume because of backward compatibility from the Xunit side. 

I tested this in a case. We had Xunit conservative mode enabled in a project with Reqnroll, all tests would still span up in parallel (e.g. `agressive` mode), even the direct async Xunit tests.

Then we removed the Reqnroll NuGet from the  (and the feature files) from the Project. So that the Reqnroll Runner was not registered in the generated code (because no code was generated), then Xunit suddenly respected the `parallelAlgorithm` setting and just run as much test in parallel as configured. 

We could easily see this, because each test created a browser instance (via Playwright). With Regnroll with had dozens of open browser (which leads to resource problems), without Reqnroll, it opens just 4 (our max threads) in `conservative` mode.

To fix this Issue, the Reqnroll Xunit Runner should be adapted so that the underlying implementation respects the conservative mode, don't know what to do for this. It could be just instruct Xunit to do so, or implement the algorithm by itself (which i don't expect).

**Reply by gasparnagy** [reply]
@DerAlbertCom This is a good info. Could you please create an issue for that at https://github.com/reqnroll/Reqnroll/issues/? If you can also help with the fix, that is also appreciated because you seem to have investigated this topic already pretty deeply.

**Reply by DerAlbertCom** [reply]
I asked https://github.com/xunit/xunit/discussions/3176 for a hint, hope to get a good answer to Kickstart the effort.

**Comment by DerAlbertCom** [comment]
So, for xUnit v2 i think i have now all Information, it's basically "reimplementing" the `XunitTestAssemblyRunner` class instead of inheriting it (copying the existing code - about 350 Lines), adapting it to regnroll, just `XunitTestAssemblyRunner` not the base classes of that). The xUnit V2 code is mature and will not be changed anymore. Is this something which can Reqnroll can live with? 

 xUnit v2 is done and finished. To really fix this little flaw  the `XunitTestAssemblyRunner` class would needs to be changed. I would not expect any change for this.
 
 For xUnit 3 i have not looked, but it's seems way different and needs to solve differently, and also we have the offer hat xunit v3 could improved on an API if we have to much work to integrate it.

**Reply by gasparnagy** [reply]
Thx. I think this fine (take in the code). I have also commented the related issue: https://github.com/reqnroll/Reqnroll/issues/473
